### PR TITLE
Conan package distribution

### PR DESCRIPTION
**Pull request by uilianries**
Hi!

This PR is related to support BehaviorTree.CPP as distributed package.

Also, including:

- conan/build.py
  Build script to create Conan package on Travis CI
- conan/test_package/CMakeLists.txt
  CMake file to build test package
- conan/test_package/conanfile.py
  Conan recipe to build test package
- conan/test_package/test_package.cpp
  Example code to consume Conan package created
- .travis.yml
  Update to run Conan job on CI

Conan uses the file conanfile.py as recipe to execute all steps necessary to create the package. As your project is a library, it needs to be built, and Conan defines the package id by settings + options + requires.

To make sure that your package is working well, Conan provides test_package, where we put a simple example consuming your package and building as well. When test_package.cpp is including BehaviorTree header, all files are imported from the packaged provided by Conan. Test package is not for unit testing propose, it's a package validation.

When running your package on CI (appveyor, travis or any other) we could run Conan to create the package and check. But prepare a full environment with compiler, build-essential , conan and all stuff could be boring, so we use Docker. Conan provides a good range of images, including gcc and clang. I added multiple jobs because your project should match (compiler version, arch, os, release/debug) when required as dependency.

#### How to upload?

Conan uses Bintray to store their packages. You need to create an account on bintray for that. There is a good documentation how to create a bintray account [here](https://docs.conan.io/en/latest/uploading_packages/bintray/uploading_bintray.html).

In short, you need to create a new account [here](https://bintray.com/signup/oss). 
So you will have two options:
  - Create a repository based on your username
  - Create a new organization and so create a new repository.

I would recommend the second one, because you will be able to add more people to maintain your organization.

Following the second choice, create a new organization named _BehaviorTree_ or any preferred name. And create a new repository named _conan_ or any preferred name.

The build.py is configured following the reference BehaviorTree/conan/BehaviorTree.CPP/. Where:

    BehaviorTree is your Bintray organization name
    conan is the repository name
    BehaviorTree.CPP is the package name

Also, build.py has a rule to upload only tags or any branch with the pattern name "\d+.\d+.\d+".

And finally, you must set the variable CONAN_PASSWORD on Travis settings. The password is the Bintray API Key, it's not your account password. To read your API key you need to access https://bintray.com/profile/edit and click on API Key button. Don't forget to mark as secret on Travis settings.

#### Bintray package page

After to upload your package by CI, we recommend update your Bintray package page:
https://bintray.com/BehaviorTree/conan/BehaviorTree.CPP%3ABehaviorTree/edit?tab=general
The page contains only project metadata, as url, maturity (official), license.

#### Conan Center request

Now that all steps are done. Just click on "Add to Conan Center" button, present on your package page, and wait for a review.

Okay, this is a lot of information. Please, feel free to ask anything.

I'm including @danimtb and @SSE4 to review this PR. They are Conan experts.

Regards!

/cc @danimtb @SSE4

related issue #2

### CODE REVIEW

File: conanfile.py [hunk 1]
```diff
@@ -41,6 +48,10 @@ def build(self):
                               """project(behaviortree_cpp)
                               include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
                               conan_basic_setup()""")
+        # INFO (uilian): zmq could requires libsodium
```
Comment (SSE4) [inline conanfile.py]: could require

File: conan/travis/install.sh [hunk 2]
```diff
@@ -0,0 +1,22 @@
+#!/bin/bash
+
+set -e
```
Comment (SSE4) [inline conan/travis/install.sh]: you may just write set -ex

File: conan/test_package/test_package.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
+#include "behaviortree_cpp/behavior_tree.h"
+#include "behaviortree_cpp/bt_factory.h"
+
+using namespace BT;
+
+NodeStatus SayHello()
+{
+    printf("hello\n");
```
Comment (SSE4) [inline conan/test_package/test_package.cpp]: why do you mix printf and cout in this example?

File: conan/test_package/test_package.cpp [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    }
+
+  private:
+    int time_;
+    std::atomic_bool stop_loop_;
+};
+
+int main()
+{
+    BT::SequenceNode root("root");
```
Comment (SSE4) [inline conan/test_package/test_package.cpp]: you wrote using namespace BT, so explicit namespace can be omitted here

File: conan/test_package/test_package.cpp (hunk 3 above)
Comment (uilianries) [reply conan/test_package/test_package.cpp]: it's from https://github.com/BehaviorTree/BehaviorTree.CPP/blob/master/examples/broken_sequence.cpp

File: conan/test_package/test_package.cpp (hunk 4 above)
Comment (uilianries) [reply conan/test_package/test_package.cpp]: it's from https://github.com/BehaviorTree/BehaviorTree.CPP/blob/master/examples/broken_sequence.cpp

File: conan/travis/install.sh (hunk 2 above)
Comment (uilianries) [reply conan/travis/install.sh]: done

**Comment by uilianries** [comment]
Hi @facontidavide !

Is there anything I could do to help you in this PR?

File: conanfile.py [hunk 5]
```diff
@@ -41,6 +48,10 @@ def build(self):
                               """project(behaviortree_cpp)
                               include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
                               conan_basic_setup()""")
+        # INFO (uilian): zmq could require libsodium
```
Comment (danimtb) [inline conanfile.py]: Is this a "to-do" of this recipe or of the zmq one?

**Comment by facontidavide** [comment]
I have little or no time to review it for the time being, and my little understanding of Conan doesn-t help.

I will just merge it. Better to ask for forgiveness than permission...

**Comment by danimtb** [comment]
Thanks @facontidavide! We will keep an eye on it but just ping us if you have any trouble 😄

**Comment by facontidavide** [comment]
travis is unhappy  https://travis-ci.org/BehaviorTree/BehaviorTree.CPP/builds/473857934

image

**Comment by uilianries** [comment]
It requires https://github.com/conan-io/conan-package-tools/pull/313 but until to be merged (maybe 2 Jan), I'll provide a patch for you.

**Pull request merged**